### PR TITLE
[PyTorch] DLC optimized for Intel CPU for inference - Pytorch 2.0 and 2.1

### PR DESCRIPTION
**Pull request by aalbersk**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
Based on closed PR: https://github.com/aws/deep-learning-containers/pull/3348 These changes add support for DLC with Intel optimized software stack for inference with PyTorch 2.0 and 2.1 on Intel CPUs. Newly added DLC allows users to optimize their workloads using [Intel® Extension for PyTorch](https://github.com/intel/intel-extension-for-pytorch#intel-extension-for-pytorch). Optimizations take advantage of AVX-512 Vector Neural Network Instructions (AVX512 VNNI) and Intel® Advanced Matrix Extensions (Intel® AMX) on Intel CPUs.

Below you can see attached average latencies for Pytorch 2.0 DLC for 2 models: Densenet161 and Bert-base-uncased. Benchmark numbers for Pytorch 2.1 will be added later.

**Densenet**
| Instance Name | DLC without Intel optimizations(fp32) [s] | DLC with Intel optimizations(fp32) [s] |
| -------- | ------- | -------- |
| c7i.4xlarge | 0.125 | 0.0429 (291.38% improvement) |
| c7i.8xlarge | 0.1252 | 0.0304 (411.84% improvement) |
| c7i.16xlarge | 0.1277 | 0.0265 (481.89% improvement) |

**Bert**
| Instance Name | DLC without Intel optimizations(fp32) [s] | DLC with Intel optimizations(fp32) [s] | DLC with Intel optimizations(bf16) [s] | 
| -------- | ------- | -------- | ------- |
| c7i.4xlarge | 0.0587 | 0.0438 (134.02% improvement) | 0.0293 (200.34% improvement) |
| c7i.8xlarge | 0.0343 | 0.0253 (135.57% improvement) | 0.0206 (166.50% improvement) |
| c7i.16xlarge | 0.0314 | 0.0202 (155.47% improvement) | 0.0156 (201.28% improvement) |

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [x] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by lxning — COMMENTED** [review]
There are lots of overlapping with existing CPU DLC. It is difficult to maintain in the future. Is it possible to create the image for AMX based on the existing CPU image?

**Comment by aalbersk** [comment]
@lxning Thank you for your comments!


> There are lots of overlapping with existing CPU DLC. It is difficult to maintain in the future. Is it possible to create the image for AMX based on the existing CPU image?

Yes, I could create, for example, two targets in each Docker.cpu files: `from ec2 as intel_ec2` and `from sagemaker as intel_sagemaker` and implement intel optimizations this way without redundancy. What do you think about this approach?

> DLC build pytorch binary by itself, not directly pulled from PyTorch

Unfortunately, each IPEX version is built with public version of corresponding PyTorch version. Installing IPEX wheel with DLC build pytorch binary will fail. Why does DLC use DLC build pytorch binaries?

**Comment by aalbersk** [comment]
@lxning Are there any more questions you'd like me to answer or should I provide you any more information?
